packages/instance-model/lib/model.js

```diff
@@ -56,21 +90,145 @@ const InstanceModel = AmpersandModel.extend({
   collections: {
     databases: MongoDbDatabaseCollection,
   },
+
+  initialize() {
+    const toggle = (id, status) => {
+      if (isLoading(status)) {
+        this.set({ loadingModels: this.loadingModels.concat(id) });
+      } else {
+        this.set({
+          loadingModels: this.loadingModels.filter((modelId) => modelId !== id),
+        });
+      }
+    };
+
+    this.on('change:databases.collectionsStatus', (model, status) => {
+      // Collections don't have their own cid
+      toggle(`${model.cid}$$coll`, status);
+    });
+
+    this.on('change:databases.status', (model, status) => {
+      toggle(model.cid, status);
+    });
+
+    this.on('change:collections.status', (model, status) => {
+      toggle(model.cid, status);
+    });
+  },
+
   /**
    * @param {{ dataService: import('mongodb-data-service').DataService }} dataService
-   * @returns
+   * @returns {Promise<void>}
    */
   async fetch({ dataService }) {
     const newStatus = this.status === 'initial' ? 'fetching' : 'refreshing';
     this.set({ status: newStatus });
     try {
       const instanceInfo = await dataService.instance();
-      this.set({ status: 'ready', ...instanceInfo })
+      this.set({ status: 'ready', statusError: null, ...instanceInfo });
     } catch (err) {
-      this.set({ status: 'error' });
+      this.set({ status: 'error', statusError: err.message });
       throw err;
     }
   },
+
+  /**
+   * @param {{ dataService: import('mongodb-data-service').DataService }} dataService
+   * @returns {Promise<void>}
+   */
+  async fetchDatabases({ dataService }) {
+    const newStatus =
+      this.databasesStatus === 'initial' ? 'fetching' : 'refreshing';
+    this.set({ databasesStatus: newStatus });
+    try {
+      await this.databases.fetch({ dataService });
+      this.set({ databasesStatus: 'ready', databasesStatusError: null });
+    } catch (err) {
+      this.set({ databasesStatus: 'error', databasesStatusError: err.message });
+      throw err;
+    }
+  },
+
+  async refresh({
```

Comment:
this function, and all the other fetch functions in the model, are both stateful and async at the same time which may cause funny behaviours. We probably could avoid multiple concurrent calls to modify state with some kind of guard

Reply:
Makes sense, yeah! For context this kind of stuff already happens in Compass in a lot of places so I wasn't sure if I want to expand the scope and introduce even more changes in the models, but now that most of this fetching logic is in one place, we might as well do it 👍

Reply:
Done in 8ddc00d, but now I'm starting to think if I should actually move all these models in one place due to all the code duplication between the three, any thoughts?